### PR TITLE
Simplify neural blocks

### PR DESCRIPTION
**Pull request by ggaspersic**
Overview:
* removal of unused (non-BLAS) related functionality
* replacing weights: Vec<Weight> with weights: Vec<f32> (no need for transmute). Impact of this that the reading also needed to be changed.
* creation of `internal_forward` to generalize the matrix multiplication step

### CODE REVIEW

File: src/block_neural.rs [hunk 1]
```diff
@@ -188,86 +188,45 @@ pub fn new_neuron_block(
...
+                self.num_inputs,
+                self.output_offset,
+                self.num_neurons,
+            );
 
-    fn get_num_output_values(&self, output: graph::OutputSlot) -> usize {
-        assert!(output.get_output_index() == 0);
-        self.num_neurons
-    }
+            // This is actually speed things up considerably.
```
Comment (SkBlaz) [inline src/block_neural.rs:207]: Typo, if cleaning up, let's fix these too

File: src/block_neural.rs [hunk 2]
```diff
@@ -282,9 +241,25 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockNeuronLayer<L> {
...
+        self.internal_forward(pb, dropout_inv);
+
+        block_helpers::forward_backward(further_blocks, fb, pb, update);
+
         unsafe {
-            let bias_offset = self.num_inputs * self.num_neurons;
-            {
+            if update && self.neuron_type == NeuronType::WeightedSum {
+                // first we need to initialize inputs to zero
+                // TODO - what to think about this buffer
```
Comment (SkBlaz) [inline src/block_neural.rs:257]: Not sure what this means

File: src/block_neural.rs [hunk 3]
```diff
@@ -293,159 +268,76 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockNeuronLayer<L> {
...
-                                    .optimizer_data,
-                            );
-                            self.weights.get_unchecked_mut(bias_offset + j).weight -= update;
-                        }
-
-                        if self.max_norm != 0.0 && fb.example_number % 10 == 0 {
-                            let mut wsquaredsum = 0.000001; // Epsilon
+                        let norm = wsquaredsum.sqrt();
+                        if norm > self.max_norm {
+                            let scaling = self.max_norm / norm;
```
Comment (SkBlaz) [inline src/block_neural.rs:316]: if there are any divisions in "hot" parts remaining, perhaps worth considering mult+shift transforms, as e.g., https://docs.rs/strength_reduce/latest/strength_reduce/

File: src/block_neural.rs [hunk 4]
```diff
@@ -456,55 +348,87 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockNeuronLayer<L> {
...
-        } // unsafe end
+            InitType::Hu => {
+                let normal =
+                    Normal::new(0.0, (2.0 / self.num_inputs as f64).sqrt() as f64).unwrap();
+
+                for i in 0..self.bias_offset {
+                    self.weights[i as usize] = normal.sample(&mut self.rng) as f32;
+                }
+            }
+            //            InitType::RandomFirst1 => { for i in 0..self.num_inputs { self.weights[i as usize] = 1.0}},
```
Comment (SkBlaz) [inline src/block_neural.rs:399]: Redundant comments?

File: src/block_neural.rs [hunk 5]
```diff
@@ -549,7 +483,7 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockNeuronLayer<L> {
         index: usize,
         w: &[f32],
     ) -> Result<(), Box<dyn Error>> {
-        self.weights[index].weight = w[0];
+        self.weights[index] = w[0];
```
Comment (SkBlaz) [inline src/block_neural.rs:486]: Should there be tests for this struct-of-arrays rewrite perhaps also here?

File: src/block_neural.rs (hunk 2 above)
Comment (ggaspersic) [reply src/block_neural.rs:257]: That's a leftover from before. I will remove it as the comment doesn't add any value.

File: src/block_neural.rs (hunk 1 above)
Comment (ggaspersic) [reply src/block_neural.rs:207]: Removed the comment, doesn't add any context

File: src/block_neural.rs [hunk 6]
```diff
@@ -18,14 +16,13 @@ use crate::model_instance;
 use crate::optimizer;
 use crate::port_buffer;
 use crate::regressor;
-use block_helpers::{OptimizerData, Weight};
```
[the remark this answers is not in the document]
Comment (ggaspersic) [reply src/block_neural.rs:21]: Missed that, and removed it.

**Review by SkBlaz — APPROVED** [review]

**Pull request merged**